{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Presenter notes with Carpentries Python Gapminder training\n",
    "*Martijn Wehrens, m.wehrens@uva.nl, 2025-03-04*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Lesson 8, Pandas dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make sure they have\n",
    "    # Gapminder in ./data folder\n",
    "        # https://swcarpentry.github.io/python-novice-gapminder/ (python-novice-gapminder-data.zip)\n",
    "    # Kohela in ./data folder\n",
    "        # https://www.biodsc.nl/workshop-materials/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1601.056136\n",
      "             gdpPercap_1962  gdpPercap_1967  gdpPercap_1972\n",
      "country                                                    \n",
      "Italy           8243.582340    10022.401310    12269.273780\n",
      "Montenegro      4649.593785     5907.850937     7778.414017\n",
      "Netherlands    12790.849560    15363.251360    18794.745670\n",
      "Norway         13450.401510    16361.876470    18965.055510\n",
      "Poland          5338.752143     6557.152776     8006.506993\n",
      "gdpPercap_1962    13450.40151\n",
      "gdpPercap_1967    16361.87647\n",
      "gdpPercap_1972    18965.05551\n",
      "dtype: float64\n"
     ]
    }
   ],
   "source": [
    "# Pandas dataframe\n",
    "    # Collection of \"series\"   \n",
    "        # one series is one column\n",
    "        # on which you can perform operations\n",
    "    # Built on \"numpy\", \n",
    "        # a library for working with arrays and mathematical operations\n",
    "        # series very similar to numpy.array\n",
    "    # Advantages\n",
    "        # Access to individual records\n",
    "        # Convenient ways combining information multiple dataframes\n",
    "        \n",
    "# how to access dataframe\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# data = pd.read_csv('../data/gapminder_gdp_europe.csv', index_col='country')\n",
    "data = pd.read_csv('/Users/m.wehrens/Data_UVA/2024_teaching/2025-03-gapminder/data/gapminder_gdp_europe.csv', \\\n",
    "    index_col='country')\n",
    "\n",
    "# print(data.head())\n",
    "\n",
    "# TWO METHODS TO ACCESS DATA\n",
    "    # iloc\n",
    "    # loc\n",
    "\n",
    "# \"iloc\" method\n",
    "# numerical indexing of rows and columns\n",
    "    # remember, it's a 2D table\n",
    "    # indexing similar as accessing chars in string\n",
    "\n",
    "print(data.iloc[0, 0])\n",
    "\n",
    "# \"loc\" method\n",
    "# acess entries by their label\n",
    "\n",
    "# Other synthax\n",
    "    # \":\" means all columns, or all rows\n",
    "    \n",
    "# print(data.loc[\"Albania\", :])\n",
    "# print(data.loc[\"Albania\"])\n",
    "    # in loc, first entry is row\n",
    "    \n",
    "# other way around\n",
    "# print(data.loc[:, \"gdpPercap_1952\"])\n",
    "# print(data[\"gdpPercap_1952\"])\n",
    "    # no \"loc\"!!!\n",
    "    # <--> series\n",
    "    \n",
    "# also works:\n",
    "# data.gdpPercap_1992\n",
    "    # not recommended, confusion with methods\n",
    "    \n",
    "# slicing works on labels as well\n",
    "print(data.loc['Italy':'Poland', 'gdpPercap_1962':'gdpPercap_1972'])\n",
    "\n",
    "# can be used to perform operations on subsets\n",
    "print(data.loc['Italy':'Poland', 'gdpPercap_1962':'gdpPercap_1972'].max())\n",
    "    # note: applied per column\n",
    "    \n",
    "\n",
    "         \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Subset of data:\n",
      "              gdpPercap_1962  gdpPercap_1967  gdpPercap_1972\n",
      "country                                                    \n",
      "Italy           8243.582340    10022.401310    12269.273780\n",
      "Montenegro      4649.593785     5907.850937     7778.414017\n",
      "Netherlands    12790.849560    15363.251360    18794.745670\n",
      "Norway         13450.401510    16361.876470    18965.055510\n",
      "Poland          5338.752143     6557.152776     8006.506993\n",
      "\n",
      "Where are values large?\n",
      "              gdpPercap_1962  gdpPercap_1967  gdpPercap_1972\n",
      "country                                                    \n",
      "Italy                 False            True            True\n",
      "Montenegro            False           False           False\n",
      "Netherlands            True            True            True\n",
      "Norway                 True            True            True\n",
      "Poland                False           False           False\n"
     ]
    }
   ],
   "source": [
    "# selecting data\n",
    "\n",
    "\n",
    "# Use a subset of data to keep output readable.\n",
    "subset = data.loc['Italy':'Poland', 'gdpPercap_1962':'gdpPercap_1972']\n",
    "print('Subset of data:\\n', subset)\n",
    "    # notice the escape \"\\\"; \\n is signifier for nextline\n",
    "\n",
    "# Which values were greater than 10000 ?\n",
    "print('\\nWhere are values large?\\n', subset > 10000)\n",
    "    # comparisons are done element-wise\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             gdpPercap_1962  gdpPercap_1967  gdpPercap_1972\n",
      "country                                                    \n",
      "Italy                 False            True            True\n",
      "Montenegro            False           False           False\n",
      "Netherlands            True            True            True\n",
      "Norway                 True            True            True\n",
      "Poland                False           False           False\n",
      "             gdpPercap_1962  gdpPercap_1967  gdpPercap_1972\n",
      "country                                                    \n",
      "Italy                   NaN     10022.40131     12269.27378\n",
      "Montenegro              NaN             NaN             NaN\n",
      "Netherlands     12790.84956     15363.25136     18794.74567\n",
      "Norway          13450.40151     16361.87647     18965.05551\n",
      "Poland                  NaN             NaN             NaN\n"
     ]
    }
   ],
   "source": [
    "# a frame full of \"booleans\" (true/false)\n",
    "    # sometimes called mask\n",
    "    # (MW: Â±filter)\n",
    "\n",
    "mask = subset > 10000\n",
    "print(mask)\n",
    "print(subset[mask])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       gdpPercap_1962  gdpPercap_1967  gdpPercap_1972\n",
      "count        2.000000        3.000000        3.000000\n",
      "mean     13120.625535    13915.843047    16676.358320\n",
      "std        466.373656     3408.589070     3817.597015\n",
      "min      12790.849560    10022.401310    12269.273780\n",
      "25%      12955.737548    12692.826335    15532.009725\n",
      "50%      13120.625535    15363.251360    18794.745670\n",
      "75%      13285.513522    15862.563915    18879.900590\n",
      "max      13450.401510    16361.876470    18965.055510\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# returns value where True\n",
    "# returns NaN where False\n",
    "    # NaN = Not a number\n",
    "    # NaNs are ignored in operations\n",
    "        # e.g. min, max, etc\n",
    "\n",
    "# e.g.\n",
    "print(subset[subset > 10000].describe())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group By: split-apply-combine\n",
    "\n",
    "# (REMOVED)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                A1BG      A1CF    A4GALT      AAAS    AACSP1      AACS  \\\n",
      "Celltype                                                                 \n",
      "epicardial  0.428571  0.080000  0.137143  2.920000  0.205714  2.137143   \n",
      "fat         0.081081  0.135135  0.405405  3.837838  0.108108  1.891892   \n",
      "fibroblast  0.348485  0.121212  0.348485  5.257576  0.621212  1.848485   \n",
      "unknown     0.156701  0.084536  0.123711  2.122680  0.108247  1.181443   \n",
      "\n",
      "               AADAT     AAED1     AAGAB      AAK1  ...     ZUFSP      ZW10  \\\n",
      "Celltype                                            ...                       \n",
      "epicardial  0.388571  0.320000  1.234286  3.291429  ...  0.222857  0.662857   \n",
      "fat         1.621622  0.486486  2.945946  7.351351  ...  0.621622  1.243243   \n",
      "fibroblast  1.696970  0.287879  2.590909  4.000000  ...  0.575758  1.393939   \n",
      "unknown     0.613402  0.213402  1.042268  2.567010  ...  0.250515  0.535052   \n",
      "\n",
      "              ZWILCH     ZWINT      ZXDB      ZXDC    ZYG11B        ZYX  \\\n",
      "Celltype                                                                  \n",
      "epicardial  2.028571  3.022857  0.451429  1.897143  3.148571  24.217143   \n",
      "fat         4.729730  3.540541  0.351351  2.756757  2.972973  26.972973   \n",
      "fibroblast  4.696970  5.696970  0.545455  3.560606  3.318182  23.015152   \n",
      "unknown     1.772165  2.101031  0.300000  1.453608  1.947423  14.983505   \n",
      "\n",
      "               ZZEF1      ZZZ3  \n",
      "Celltype                        \n",
      "epicardial  1.022857  1.154286  \n",
      "fat         1.459459  1.945946  \n",
      "fibroblast  0.909091  1.984848  \n",
      "unknown     0.709278  0.858763  \n",
      "\n",
      "[4 rows x 12237 columns]\n",
      "A1BG\n",
      "ZZZ3\n",
      "Celltype\n",
      "epicardial    0.223389\n",
      "fat           2.620107\n",
      "fibroblast    3.229871\n",
      "unknown       0.926589\n",
      "Name: TFAP2A, dtype: float64\n",
      "             ADAMTS19     COL2A1    CYP26A1        GAL      HHLA2      PPARG  \\\n",
      "Celltype                                                                       \n",
      "epicardial   0.059429   0.092746   0.021099   0.162078   0.144069   0.100161   \n",
      "fat          0.393514   0.106702   0.049896  16.481572  11.711712  14.022472   \n",
      "fibroblast  10.809697  11.411919  10.125874   2.256198   0.740129   0.106231   \n",
      "unknown      0.525361   0.489314   0.591911   0.475164   0.763511   0.726422   \n",
      "\n",
      "                SALL1   SERPINB2    SPARCL1     SPTSSB      SUSD2       TAC3  \\\n",
      "Celltype                                                                       \n",
      "epicardial   0.070608   0.255836   0.008326   0.057512   0.443733   0.163066   \n",
      "fat          0.222638  13.461641   2.303782  11.968614  10.493694  11.603164   \n",
      "fibroblast  10.733873   0.000000  10.056148   0.304985   0.336162   0.451589   \n",
      "unknown      0.535021   0.726957   0.512987   0.798936   0.783395   0.783857   \n",
      "\n",
      "              TMEM139  \n",
      "Celltype               \n",
      "epicardial   0.194494  \n",
      "fat         10.118919  \n",
      "fibroblast   0.515702  \n",
      "unknown      0.830440  \n"
     ]
    }
   ],
   "source": [
    "# Group by: split-apply-combine\n",
    "    # note to self: see related/202503_RNAseq-data.py\n",
    "\n",
    "# Goal\n",
    "    # table in which rows match condition\n",
    "    # calculate something per condition\n",
    "    # use group_by\n",
    "    # Using data from Kohela et al. \n",
    "        # RNA-seq data (powerpoint)\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# Import data\n",
    "df_cells_kohela = pd.read_csv('/Users/m.wehrens/Data_UVA/example-datasets/kohela-et-al/kohela-et-al.csv', index_col=0)\n",
    "df_cells_kohela2 = df_cells_kohela.T\n",
    "df_cells_kohela2.head()\n",
    "\n",
    "# create new 'masks'\n",
    "epicardial_cells = df_cells_kohela2['WT1']>3\n",
    "fibroblast_cells = df_cells_kohela2['COL2A1']>30\n",
    "fat_cells = df_cells_kohela2['PPARG']>2\n",
    "\n",
    "# Create a new column\n",
    "df_cells_kohela2['Celltype'] = 'unknown'\n",
    "# Set values for the column\n",
    "df_cells_kohela2.loc[epicardial_cells,'Celltype'] = 'epicardial'\n",
    "df_cells_kohela2.loc[fibroblast_cells, 'Celltype'] = 'fibroblast'\n",
    "df_cells_kohela2.loc[fat_cells, 'Celltype'] = 'fat'\n",
    "\n",
    "# OPTIONAL: give an overview of the frequencies of 'Celltype' values\n",
    "df_cells_kohela2['Celltype'].value_counts()\n",
    "\n",
    "# now use group_by to calculate gene expression median values per group\n",
    "    # explain:\n",
    "        # creates grouped object (groupby object)\n",
    "        # now any aggregate function \n",
    "            # e.g. aggregate('XX'), max(), min(), etc\n",
    "            # applied per group\n",
    "            # will paste together the dataframe\n",
    "df_cells_kohela2_groupedType = df_cells_kohela2.groupby('Celltype')\n",
    "\n",
    "# Now calculate mean expression\n",
    "df_results = df_cells_kohela2_groupedType.mean() \n",
    "print(df_results.head())\n",
    "\n",
    "print(df_cells_kohela2.columns[0])\n",
    "print(df_cells_kohela2.columns[-2])\n",
    "\n",
    "df_means = df_cells_kohela2.loc[:,'A1BG':'ZZZ3'].mean()\n",
    "\n",
    "# And normalize\n",
    "df_results_normalized = df_results/df_means\n",
    "\n",
    "# \"Check\" their TFAP2A claim\n",
    "print(df_results_normalized.loc[:,'TFAP2A'])\n",
    "\n",
    "# Check some other highly differentially expressed genes\n",
    "print(df_results_normalized.loc[:,(df_results_normalized>10).any()])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extra solution stuff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['gdpPercap_1952', 'gdpPercap_1957', 'gdpPercap_1962', 'gdpPercap_1967',\n",
      "       'gdpPercap_1972', 'gdpPercap_1977', 'gdpPercap_1982', 'gdpPercap_1987',\n",
      "       'gdpPercap_1992', 'gdpPercap_1997', 'gdpPercap_2002', 'gdpPercap_2007'],\n",
      "      dtype='object')\n",
      "Index(['Albania', 'Austria', 'Belgium', 'Bosnia and Herzegovina', 'Bulgaria',\n",
      "       'Croatia', 'Czech Republic', 'Denmark', 'Finland', 'France', 'Germany',\n",
      "       'Greece', 'Hungary', 'Iceland', 'Ireland', 'Italy', 'Montenegro',\n",
      "       'Netherlands', 'Norway', 'Poland', 'Portugal', 'Romania', 'Serbia',\n",
      "       'Slovak Republic', 'Slovenia', 'Spain', 'Sweden', 'Switzerland',\n",
      "       'Turkey', 'United Kingdom'],\n",
      "      dtype='object', name='country')\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "gdpPercap_1952    27\n",
       "gdpPercap_1957    27\n",
       "gdpPercap_1962    27\n",
       "gdpPercap_1967    27\n",
       "gdpPercap_1972    27\n",
       "gdpPercap_1977    27\n",
       "gdpPercap_1982    27\n",
       "gdpPercap_1987    18\n",
       "gdpPercap_1992    18\n",
       "gdpPercap_1997    18\n",
       "gdpPercap_2002    18\n",
       "gdpPercap_2007    18\n",
       "dtype: int64"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Some convenient stuff for solutions\n",
    "# Only selected exercises\n",
    "\n",
    "###\n",
    "# Extent of slicing\n",
    "\n",
    "#data_europe = pd.read_csv('data/gapminder_gdp_europe.csv', index_col='country')\n",
    "data_europe = pd.read_csv('/Users/m.wehrens/Data_UVA/2024_teaching/2025-03-gapminder/data/gapminder_gdp_europe.csv', \\\n",
    "    index_col='country')\n",
    "\n",
    "print(data_europe.columns) # print col names\n",
    "print(data_europe.index) # print row names\n",
    "\n",
    "###\n",
    "# Selecting Indices, how to get indices\n",
    "\n",
    "# option 1\n",
    "import numpy as np\n",
    "data_europe.aggregate(np.argmax)\n",
    "# option 2\n",
    "data_europe.reset_index(drop=True).idxmax()\n",
    "# option 3\n",
    "# ..? \n",
    "# not sure the others are elegant..\n",
    "\n",
    "###\n",
    "#  Practice with Selection\n",
    "\n",
    "# - GDP per capita for all countries in 1982.\n",
    "# rows\n",
    "print(data_europe.index)\n",
    "# cols:\n",
    "print(data_europe.T.index)\n",
    "print(data_europe.keys())\n",
    "print(data_europe.columns)\n",
    "# Solution:\n",
    "# data_europe.loc[:,'gdpPercap_1982']\n",
    "\n",
    "# - GDP per capita for Denmark for all years.\n",
    "# data_europe.loc['Denmark',:]\n",
    "\n",
    "# - GDP per capita for all countries for years after 1985.\n",
    "# data_europe.loc[:, 'gdpPercap_1982':]\n",
    "\n",
    "# - GDP per capita for each country in 2007 as a multiple of GDP per capita for that country in 1952.\n",
    "# GDP_increase_factor = data_europe.loc[:, 'gdpPercap_2007'] / data_europe.loc[:, 'gdpPercap_1952']\n",
    "# print(GDP_increase_factor)\n",
    "\n",
    "###\n",
    "# Many Ways of Access\n",
    "\n",
    "# See online solution at https://swcarpentry.github.io/python-novice-gapminder/instructor/08-data-frames.html\n",
    "# (CTRL+F: \"Many Ways of Access\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "2025_IMG",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
