{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Presenter notes with Carpentries Python Gapminder training\n",
    "*Martijn Wehrens, m.wehrens@uva.nl, 2025-03-04*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Lesson 8, Pandas dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1601.056136\n",
      "             gdpPercap_1962  gdpPercap_1967  gdpPercap_1972\n",
      "country                                                    \n",
      "Italy           8243.582340    10022.401310    12269.273780\n",
      "Montenegro      4649.593785     5907.850937     7778.414017\n",
      "Netherlands    12790.849560    15363.251360    18794.745670\n",
      "Norway         13450.401510    16361.876470    18965.055510\n",
      "Poland          5338.752143     6557.152776     8006.506993\n",
      "gdpPercap_1962    13450.40151\n",
      "gdpPercap_1967    16361.87647\n",
      "gdpPercap_1972    18965.05551\n",
      "dtype: float64\n"
     ]
    }
   ],
   "source": [
    "# Pandas dataframe\n",
    "    # Collection of \"series\"   \n",
    "        # one series is one column\n",
    "        # on which you can perform operations\n",
    "    # Built on \"numpy\", \n",
    "        # a library for working with arrays and mathematical operations\n",
    "        # series very similar to numpy.array\n",
    "    # Advantages\n",
    "        # Access to individual records\n",
    "        # Convenient ways combining information multiple dataframes\n",
    "        \n",
    "# how to access dataframe\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# data = pd.read_csv('../data/gapminder_gdp_europe.csv', index_col='country')\n",
    "data = pd.read_csv('/Users/m.wehrens/Data_UVA/2024_teaching/2025-03-gapminder/data/gapminder_gdp_europe.csv', \\\n",
    "    index_col='country')\n",
    "\n",
    "# print(data.head())\n",
    "\n",
    "# TWO METHODS TO ACCESS DATA\n",
    "    # iloc\n",
    "    # loc\n",
    "\n",
    "# \"iloc\" method\n",
    "# numerical indexing of rows and columns\n",
    "    # remember, it's a 2D table\n",
    "    # indexing similar as accessing chars in string\n",
    "\n",
    "print(data.iloc[0, 0])\n",
    "\n",
    "# \"loc\" method\n",
    "# acess entries by their label\n",
    "\n",
    "# Other synthax\n",
    "    # \":\" means all columns, or all rows\n",
    "    \n",
    "# print(data.loc[\"Albania\", :])\n",
    "# print(data.loc[\"Albania\"])\n",
    "    # in loc, first entry is row\n",
    "    \n",
    "# other way around\n",
    "# print(data.loc[:, \"gdpPercap_1952\"])\n",
    "# print(data[\"gdpPercap_1952\"])\n",
    "    # no \"loc\"!!!\n",
    "    # <--> series\n",
    "    \n",
    "# also works:\n",
    "# data.gdpPercap_1992\n",
    "    # not recommended, confusion with methods\n",
    "    \n",
    "# slicing works on labels as well\n",
    "print(data.loc['Italy':'Poland', 'gdpPercap_1962':'gdpPercap_1972'])\n",
    "\n",
    "# can be used to perform operations on subsets\n",
    "print(data.loc['Italy':'Poland', 'gdpPercap_1962':'gdpPercap_1972'].max())\n",
    "    # note: applied per column\n",
    "    \n",
    "\n",
    "         \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Subset of data:\n",
      "              gdpPercap_1962  gdpPercap_1967  gdpPercap_1972\n",
      "country                                                    \n",
      "Italy           8243.582340    10022.401310    12269.273780\n",
      "Montenegro      4649.593785     5907.850937     7778.414017\n",
      "Netherlands    12790.849560    15363.251360    18794.745670\n",
      "Norway         13450.401510    16361.876470    18965.055510\n",
      "Poland          5338.752143     6557.152776     8006.506993\n",
      "\n",
      "Where are values large?\n",
      "              gdpPercap_1962  gdpPercap_1967  gdpPercap_1972\n",
      "country                                                    \n",
      "Italy                 False            True            True\n",
      "Montenegro            False           False           False\n",
      "Netherlands            True            True            True\n",
      "Norway                 True            True            True\n",
      "Poland                False           False           False\n"
     ]
    }
   ],
   "source": [
    "# selecting data\n",
    "\n",
    "\n",
    "# Use a subset of data to keep output readable.\n",
    "subset = data.loc['Italy':'Poland', 'gdpPercap_1962':'gdpPercap_1972']\n",
    "print('Subset of data:\\n', subset)\n",
    "    # notice the escape \"\\\"; \\n is signifier for nextline\n",
    "\n",
    "# Which values were greater than 10000 ?\n",
    "print('\\nWhere are values large?\\n', subset > 10000)\n",
    "    # comparisons are done element-wise\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             gdpPercap_1962  gdpPercap_1967  gdpPercap_1972\n",
      "country                                                    \n",
      "Italy                 False            True            True\n",
      "Montenegro            False           False           False\n",
      "Netherlands            True            True            True\n",
      "Norway                 True            True            True\n",
      "Poland                False           False           False\n",
      "             gdpPercap_1962  gdpPercap_1967  gdpPercap_1972\n",
      "country                                                    \n",
      "Italy                   NaN     10022.40131     12269.27378\n",
      "Montenegro              NaN             NaN             NaN\n",
      "Netherlands     12790.84956     15363.25136     18794.74567\n",
      "Norway          13450.40151     16361.87647     18965.05551\n",
      "Poland                  NaN             NaN             NaN\n"
     ]
    }
   ],
   "source": [
    "# a frame full of \"booleans\" (true/false)\n",
    "    # sometimes called mask\n",
    "    # (MW: ±filter)\n",
    "\n",
    "mask = subset > 10000\n",
    "print(mask)\n",
    "print(subset[mask])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       gdpPercap_1962  gdpPercap_1967  gdpPercap_1972\n",
      "count        2.000000        3.000000        3.000000\n",
      "mean     13120.625535    13915.843047    16676.358320\n",
      "std        466.373656     3408.589070     3817.597015\n",
      "min      12790.849560    10022.401310    12269.273780\n",
      "25%      12955.737548    12692.826335    15532.009725\n",
      "50%      13120.625535    15363.251360    18794.745670\n",
      "75%      13285.513522    15862.563915    18879.900590\n",
      "max      13450.401510    16361.876470    18965.055510\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# returns value where True\n",
    "# returns NaN where False\n",
    "    # NaN = Not a number\n",
    "    # NaNs are ignored in operations\n",
    "        # e.g. min, max, etc\n",
    "\n",
    "# e.g.\n",
    "print(subset[subset > 10000].describe())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group By: split-apply-combine\n",
    "\n",
    "# (REMOVED)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                A1BG      A1CF    A4GALT      AAAS    AACSP1      AACS  \\\n",
      "Celltype                                                                 \n",
      "epicardial  0.428571  0.080000  0.137143  2.920000  0.205714  2.137143   \n",
      "fat         0.081081  0.135135  0.405405  3.837838  0.108108  1.891892   \n",
      "fibroblast  0.348485  0.121212  0.348485  5.257576  0.621212  1.848485   \n",
      "unknown     0.156701  0.084536  0.123711  2.122680  0.108247  1.181443   \n",
      "\n",
      "               AADAT     AAED1     AAGAB      AAK1  ...     ZUFSP      ZW10  \\\n",
      "Celltype                                            ...                       \n",
      "epicardial  0.388571  0.320000  1.234286  3.291429  ...  0.222857  0.662857   \n",
      "fat         1.621622  0.486486  2.945946  7.351351  ...  0.621622  1.243243   \n",
      "fibroblast  1.696970  0.287879  2.590909  4.000000  ...  0.575758  1.393939   \n",
      "unknown     0.613402  0.213402  1.042268  2.567010  ...  0.250515  0.535052   \n",
      "\n",
      "              ZWILCH     ZWINT      ZXDB      ZXDC    ZYG11B        ZYX  \\\n",
      "Celltype                                                                  \n",
      "epicardial  2.028571  3.022857  0.451429  1.897143  3.148571  24.217143   \n",
      "fat         4.729730  3.540541  0.351351  2.756757  2.972973  26.972973   \n",
      "fibroblast  4.696970  5.696970  0.545455  3.560606  3.318182  23.015152   \n",
      "unknown     1.772165  2.101031  0.300000  1.453608  1.947423  14.983505   \n",
      "\n",
      "               ZZEF1      ZZZ3  \n",
      "Celltype                        \n",
      "epicardial  1.022857  1.154286  \n",
      "fat         1.459459  1.945946  \n",
      "fibroblast  0.909091  1.984848  \n",
      "unknown     0.709278  0.858763  \n",
      "\n",
      "[4 rows x 12237 columns]\n",
      "Celltype\n",
      "epicardial     1.274286\n",
      "fat           14.945946\n",
      "fibroblast    18.424242\n",
      "unknown        5.285567\n",
      "Name: TFAP2A, dtype: float64\n",
      "                ABCG2       ACLY      ACSL3      ACTA2        ACTB      ACTC1  \\\n",
      "Celltype                                                                        \n",
      "epicardial   0.222857   6.502857   6.017143   3.205714  164.680000   0.851429   \n",
      "fat         12.675676   9.675676  20.054054  25.621622  414.864865   6.243243   \n",
      "fibroblast   5.606061  12.924242   6.530303  10.212121  264.818182  34.348485   \n",
      "unknown      1.497938   4.656701   6.378351   5.089691  160.623711   7.108247   \n",
      "\n",
      "                 ACTG1      ACTG2      ACTN1      ACTN4  ...      YWHAE  \\\n",
      "Celltype                                                 ...              \n",
      "epicardial  173.605714   4.954286  10.422857  19.371429  ...  23.234286   \n",
      "fat         254.837838  53.918919  38.648649  37.486486  ...  36.891892   \n",
      "fibroblast  179.333333   6.318182  22.530303  21.090909  ...  45.484848   \n",
      "unknown     128.282474  10.197938  13.998969  17.340206  ...  19.178351   \n",
      "\n",
      "                YWHAG      YWHAQ      YWHAZ    ZFP36L1    ZFP36L2     ZNF207  \\\n",
      "Celltype                                                                       \n",
      "epicardial   8.017143  22.525714  28.457143  10.868571   3.668571   5.937143   \n",
      "fat         10.513514  34.540541  79.459459  15.837838   5.648649   8.702703   \n",
      "fibroblast  12.939394  54.651515  56.787879  14.575758  10.166667  11.606061   \n",
      "unknown      5.929897  19.067010  32.405155   8.368041   3.353608   4.309278   \n",
      "\n",
      "               ZNF664  ZNFX1-AS1        ZYX  \n",
      "Celltype                                     \n",
      "epicardial   4.400000   8.685714  24.217143  \n",
      "fat          6.918919  12.729730  26.972973  \n",
      "fibroblast  13.212121  13.515152  23.015152  \n",
      "unknown      4.165979   6.513402  14.983505  \n",
      "\n",
      "[4 rows x 1167 columns]\n"
     ]
    }
   ],
   "source": [
    "# Group by: split-apply-combine\n",
    "    # note to self: see related/202503_RNAseq-data.py\n",
    "\n",
    "# Goal\n",
    "    # table in which rows match condition\n",
    "    # calculate something per condition\n",
    "    # use group_by\n",
    "    # Using data from Kohela et al. \n",
    "        # RNA-seq data (powerpoint)\n",
    "\n",
    "# Import data\n",
    "df_cells_kohela = pd.read_csv('/Users/m.wehrens/Data_UVA/example-datasets/kohela-et-al/kohela-et-al.csv', index_col=0)\n",
    "df_cells_kohela2 = df_cells_kohela.T\n",
    "df_cells_kohela2.head()\n",
    "\n",
    "# create new 'masks'\n",
    "epicardial_cells = df_cells_kohela2['WT1']>3\n",
    "fibroblast_cells = df_cells_kohela2['COL2A1']>30\n",
    "fat_cells = df_cells_kohela2['PPARG']>2\n",
    "\n",
    "# Create a new column\n",
    "df_cells_kohela2['Celltype'] = 'unknown'\n",
    "# Set values for the column\n",
    "df_cells_kohela2.loc[epicardial_cells,'Celltype'] = 'epicardial'\n",
    "df_cells_kohela2.loc[fibroblast_cells, 'Celltype'] = 'fibroblast'\n",
    "df_cells_kohela2.loc[fat_cells, 'Celltype'] = 'fat'\n",
    "\n",
    "# OPTIONAL: give an overview of the frequencies of 'Celltype' values\n",
    "df_cells_kohela2['Celltype'].value_counts()\n",
    "\n",
    "# now use group_by to calculate gene expression median values per group\n",
    "    # explain:\n",
    "        # assigns hidden grouping parameter\n",
    "        # now any aggregate function \n",
    "            # e.g. aggregate('XX'), max(), min(), etc\n",
    "            # applied per group\n",
    "df_cells_kohela2_groupedType = df_cells_kohela2.groupby('Celltype')\n",
    "\n",
    "# Now calculate mean expression\n",
    "df_results = df_cells_kohela2_groupedType.mean() \n",
    "print(df_results.head())\n",
    "df_results_normalized = df_results / df_cells_kohela2.iloc[:,:-1].mean()\n",
    "\n",
    "# \"Check\" their TFAP2A claim\n",
    "print(df_results.loc[:,'TFAP2A'])\n",
    "\n",
    "# Check some other highly differentially expressed genes\n",
    "print(df_results.loc[:,(df_results>10).any()])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Questions\n",
    "\n",
    "### Selection individual values\n",
    "\n",
    "Import data for europe:\n",
    "\n",
    "```\n",
    "import pandas as pd\n",
    "data_europe = pd.read_csv('data/gapminder_gdp_europe.csv', index_col='country')\n",
    "```\n",
    "\n",
    "Find the GDP Capita of Serbia in 2007.\n",
    "\n",
    "### Extent of slicing\n",
    "\n",
    "- Do the two statements below produce the same output?\n",
    "- Based on this, what rule governs what is included (or not) in numerical slices and named slices in Pandas?\n",
    "\n",
    "```python\n",
    "print(data_europe.iloc[0:2, 0:2])\n",
    "print(data_europe.loc['Albania':'Belgium', 'gdpPercap_1952':'gdpPercap_1962'])\n",
    "```\n",
    "\n",
    "### Reconstructing Data\n",
    "\n",
    "Explain what each line in the following short program does: what is in first, second, etc.?\n",
    "\n",
    "```python\n",
    "first = pd.read_csv('data/gapminder_all.csv', index_col='country')\n",
    "second = first[first['continent'] == 'Americas']\n",
    "third = second.drop('Puerto Rico')\n",
    "fourth = third.drop('continent', axis = 1)\n",
    "fourth.to_csv('result.csv')\n",
    "```\n",
    "\n",
    "### Selecting Indices\n",
    "\n",
    "Explain in simple terms what idxmin and idxmax do in the short program below. When would you use these methods?\n",
    "\n",
    "```python\n",
    "data = pd.read_csv('data/gapminder_gdp_europe.csv', index_col='country')\n",
    "print(data.idxmin())\n",
    "print(data.idxmax())\n",
    "```\n",
    "\n",
    "### Practice with Selection\n",
    "\n",
    "Assume Pandas has been imported and the Gapminder GDP data for Europe has been loaded. Write an expression to select each of the following:\n",
    "\n",
    "- GDP per capita for all countries in 1982.\n",
    "- GDP per capita for Denmark for all years.\n",
    "- GDP per capita for all countries for years after 1985.\n",
    "- GDP per capita for each country in 2007 as a multiple of GDP per capita for that country in 1952.\n",
    "\n",
    "### Many Ways of Access\n",
    "\n",
    "There are at least two ways of accessing a value or slice of a DataFrame: by name or index. However, there are many others. For example, a single column or row can be accessed either as a DataFrame or a Series object.\n",
    "\n",
    "Suggest different ways of doing the following operations on a DataFrame:\n",
    "\n",
    "- Access a single column\n",
    "- Access a single row\n",
    "- Access an individual DataFrame element\n",
    "- Access several columns\n",
    "- Access several rows\n",
    "- Access a subset of specific rows and columns\n",
    "- Access a subset of row and column ranges\n",
    "\n",
    "### Exploring available methods using the dir() function\n",
    "\n",
    "Python includes a dir() function that can be used to display all of the available methods (functions) that are built into a data object. In Episode 4, we used some methods with a string. But we can see many more are available by using dir():\n",
    "\n",
    "```python\n",
    "my_string = 'Hello world!'   # creation of a string object \n",
    "dir(my_string)\n",
    "```\n",
    "\n",
    "This command returns:\n",
    "\n",
    "```python\n",
    "['__add__',\n",
    "...\n",
    "'__subclasshook__',\n",
    "'capitalize',\n",
    "'casefold',\n",
    "'center',\n",
    "...\n",
    "'upper',\n",
    "'zfill']\n",
    "```\n",
    "\n",
    "You can use help() or Shift+Tab to get more information about what these methods do.\n",
    "\n",
    "Assume Pandas has been imported and the Gapminder GDP data for Europe has been loaded as data. Then, use dir() to find the function that prints out the median per-capita GDP across all European countries for each year that information is available.\n",
    "\n",
    "### Interpretation\n",
    "\n",
    "Poland’s borders have been stable since 1945, but changed several times in the years before then. How would you handle this if you were creating a table of GDP per capita for Poland for the entire twentieth century?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Additional exercises for fast participants\n",
    "\n",
    "### GDPs\n",
    "\n",
    "- Between '87 and '92 the GDP of most countries took a hit. Are there any countries which had a positive increase between those two years? Which ones?\n",
    "- Calculate the average GDP between all European countries per year.\n",
    "    - Normalize the dataframe by this trend.\n",
    "\n",
    "### Gene expression\n",
    "\n",
    "- Convert the data below to a csv file, import it to a pandas df, and determine the following:\n",
    "    - The average CRP gene expression per condion.\n",
    "    - The corresponding standard deviations.\n",
    "    - The log2-fold change between WT, condition A, and condition B.\n",
    "    - Do the same for ACTA1.\n",
    "    - Normalize all gene expression levels to their average respective wild type levels.\n",
    "\n",
    "```\n",
    "gene\texpression\tcondition\n",
    "CRP\t873\tWT\n",
    "CRP\t324\tWT\n",
    "CRP\t214\tWT\n",
    "CRP\t151\tWT\n",
    "CRP\t1220\tA\n",
    "CRP\t450\tA\n",
    "CRP\t300\tA\n",
    "CRP\t210\tA\n",
    "CRP\t800\tB\n",
    "CRP\t200\tB\n",
    "CRP\t200\tB\n",
    "CRP\t130\tB\n",
    "ACTA1\t7457\tWT\n",
    "ACTA1\t2342\tWT\n",
    "ACTA1\t8000\tWT\n",
    "ACTA1\t9000\tWT\n",
    "ACTA1\t6500\tA\n",
    "ACTA1\t2200\tA\n",
    "ACTA1\t7500\tA\n",
    "ACTA1\t8000\tA\n",
    "ACTA1\t1000\tB\n",
    "ACTA1\t1123\tB\n",
    "ACTA1\t3211\tB\n",
    "ACTA1\t1231\tB\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extra solution stuff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['gdpPercap_1952', 'gdpPercap_1957', 'gdpPercap_1962', 'gdpPercap_1967',\n",
      "       'gdpPercap_1972', 'gdpPercap_1977', 'gdpPercap_1982', 'gdpPercap_1987',\n",
      "       'gdpPercap_1992', 'gdpPercap_1997', 'gdpPercap_2002', 'gdpPercap_2007'],\n",
      "      dtype='object')\n",
      "Index(['Albania', 'Austria', 'Belgium', 'Bosnia and Herzegovina', 'Bulgaria',\n",
      "       'Croatia', 'Czech Republic', 'Denmark', 'Finland', 'France', 'Germany',\n",
      "       'Greece', 'Hungary', 'Iceland', 'Ireland', 'Italy', 'Montenegro',\n",
      "       'Netherlands', 'Norway', 'Poland', 'Portugal', 'Romania', 'Serbia',\n",
      "       'Slovak Republic', 'Slovenia', 'Spain', 'Sweden', 'Switzerland',\n",
      "       'Turkey', 'United Kingdom'],\n",
      "      dtype='object', name='country')\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "gdpPercap_1952    27\n",
       "gdpPercap_1957    27\n",
       "gdpPercap_1962    27\n",
       "gdpPercap_1967    27\n",
       "gdpPercap_1972    27\n",
       "gdpPercap_1977    27\n",
       "gdpPercap_1982    27\n",
       "gdpPercap_1987    18\n",
       "gdpPercap_1992    18\n",
       "gdpPercap_1997    18\n",
       "gdpPercap_2002    18\n",
       "gdpPercap_2007    18\n",
       "dtype: int64"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Some convenient stuff for solutions\n",
    "# Only selected exercises\n",
    "\n",
    "###\n",
    "# Extent of slicing\n",
    "\n",
    "#data_europe = pd.read_csv('data/gapminder_gdp_europe.csv', index_col='country')\n",
    "data_europe = pd.read_csv('/Users/m.wehrens/Data_UVA/2024_teaching/2025-03-gapminder/data/gapminder_gdp_europe.csv', \\\n",
    "    index_col='country')\n",
    "\n",
    "print(data_europe.columns) # print col names\n",
    "print(data_europe.index) # print row names\n",
    "\n",
    "###\n",
    "# Selecting Indices, how to get indices\n",
    "\n",
    "# option 1\n",
    "import numpy as np\n",
    "data_europe.aggregate(np.argmax)\n",
    "# option 2\n",
    "data_europe.reset_index(drop=True).idxmax()\n",
    "# option 3\n",
    "# ..? \n",
    "# not sure the others are elegant..\n",
    "\n",
    "###\n",
    "#  Practice with Selection\n",
    "\n",
    "# - GDP per capita for all countries in 1982.\n",
    "# rows\n",
    "print(data_europe.index)\n",
    "# cols:\n",
    "print(data_europe.T.index)\n",
    "print(data_europe.keys())\n",
    "print(data_europe.columns)\n",
    "# Solution:\n",
    "# data_europe.loc[:,'gdpPercap_1982']\n",
    "\n",
    "# - GDP per capita for Denmark for all years.\n",
    "# data_europe.loc['Denmark',:]\n",
    "\n",
    "# - GDP per capita for all countries for years after 1985.\n",
    "# data_europe.loc[:, 'gdpPercap_1982':]\n",
    "\n",
    "# - GDP per capita for each country in 2007 as a multiple of GDP per capita for that country in 1952.\n",
    "# GDP_increase_factor = data_europe.loc[:, 'gdpPercap_2007'] / data_europe.loc[:, 'gdpPercap_1952']\n",
    "# print(GDP_increase_factor)\n",
    "\n",
    "###\n",
    "# Many Ways of Access\n",
    "\n",
    "# See online solution at https://swcarpentry.github.io/python-novice-gapminder/instructor/08-data-frames.html\n",
    "# (CTRL+F: \"Many Ways of Access\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Albania', 'Austria', 'Belgium', 'Bosnia and Herzegovina', 'Bulgaria',\n",
      "       'Croatia', 'Czech Republic', 'Denmark', 'Finland', 'France', 'Germany',\n",
      "       'Greece', 'Hungary', 'Iceland', 'Ireland', 'Italy', 'Montenegro',\n",
      "       'Netherlands', 'Norway', 'Poland', 'Portugal', 'Romania', 'Serbia',\n",
      "       'Slovak Republic', 'Slovenia', 'Spain', 'Sweden', 'Switzerland',\n",
      "       'Turkey', 'United Kingdom'],\n",
      "      dtype='object', name='country')\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "2025_IMG",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
